lambdaclass/mirra_backend
# Add nearest/furthest targeting strategies

Adds Nearest/Furthest targeting strategies, configurable through a `proximities.json` file. File is read on application start and whenever `Champions.Config.import_proximity_config()` is called, which updates the application variables in memory.

# To Test

1. Change Melo Basic cooldown to a really long value in `skills.json` so that it won't clutter up your logs.
2. Run this battle:
```
{:ok, user_1} = Champions.Users.register("User1")
{:ok, user_2} = Champions.Users.register("User2")
team1 = Enum.map([1], fn slot -> GameBackend.Units.insert_unit(%{character_id: GameBackend.Units.Characters.get_character_by_name("Muflus").id, level: 1, tier: 1, rank: 1, selected: true, user_id: user_1.id, slot: slot}) |> elem(1) |> GameBackend.Repo.preload(items: :template, character: [[basic_skill: [mechanics: :apply_effects_to],ultimate_skill: [mechanics: :apply_effects_to]]]) end)
team2 = Enum.map([1, 2], fn slot -> GameBackend.Units.insert_unit(%{character_id: GameBackend.Units.Characters.get_character_by_name("Melo").id, level: 1, tier: 1, rank: 1, selected: true, user_id: user_2.id, slot: slot}) |> elem(1) |> GameBackend.Repo.preload(items: :template, character: [[basic_skill: [mechanics: :apply_effects_to],ultimate_skill: [mechanics: :apply_effects_to]]]) end)
result = Champions.Battle.Simulator.run_battle(team1, team2)
```
Between a Muflus and two target dummy Melos.
3. You should see that Muflus punches a specific Melo to death (take note of its id) and then starts hitting the other one. If you check in the battle history `result.initial_state` you'll see the id matches the Melo in slot `1`
4. Change Muflus Basic target strategy to "furthest" in `skills.json`
5. run the battle again!
`result = Champions.Battle.Simulator.run_battle(team1, team2)`
You should now see that the first Melo to get annihilated is the one in slot 2 (i.e. the furthest one!)
